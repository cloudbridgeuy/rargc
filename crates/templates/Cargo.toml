[package]
name = "templates"
version = "0.0.0"
homepage = "https://github.com/cloudbridgeuy/rargc"
description = "Template library for rargc"
autobins = false

authors.workspace = true
edition.workspace = true
license.workspace = true

[lib]
name = "templates"
path = "src/lib.rs"

[dependencies]
color-eyre = "0.6.2"
log = "0.4.19"
tera = "1"
lazy_static = "1.4.0"
serde = { version = "1.0.186", features = ["derive"] }
serde_json = "1.0.105"
include_dir = "0.7.3"

[dev-dependencies]
insta = { version = "1.31.0", features = ["yaml"] }



