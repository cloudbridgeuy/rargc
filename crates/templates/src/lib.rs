use color_eyre::eyre::Result;
use include_dir::{include_dir, Dir};
use lazy_static::lazy_static;
use serde::Serialize;
use tera::{Context, Tera};

static TEMPLATES_DIR: Dir = include_dir!("$CARGO_MANIFEST_DIR/templates");

lazy_static! {
    pub static ref TEMPLATES: Tera = {
        let mut tera = match Tera::new("templates/**/*.tera") {
            Ok(t) => t,
            Err(e) => {
                log::error!("Parsing error(s): {}", e);
                ::std::process::exit(1);
            }
        };

        for entry in TEMPLATES_DIR.files() {
            match tera.add_raw_template(
                entry.path().to_str().unwrap(),
                entry.contents_utf8().unwrap(),
            ) {
                Ok(_) => (),
                Err(e) => {
                    log::error!("Parsing error(s): {}", e);
                    ::std::process::exit(1);
                }
            }
        }

        tera
    };
}

/// Render the main.sh template
pub fn render<T: Serialize>(context: &T) -> Result<String> {
    let output = TEMPLATES.render("main.tera", &Context::from_serialize(context)?)?;
    Ok(output)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_usage_sh() {
        let objects = vec![
            Context::from_serialize(serde_json::json!({
                "meta": {
                    "name": "usage",
                    "description": "Test simple usage"
                },
            })),
            Context::from_serialize(serde_json::json!({
                "meta": {
                    "name": "usage",
                    "description": "Test simple usage"
                },
                "options": {
                    "foo": {
                        "short": "f",
                        "name": "foo",
                        "summary": "Test foo option"
                    }
                }
            })),
            Context::from_serialize(serde_json::json!({
                "meta": {
                    "name": "usage",
                    "description": "Test simple usage"
                },
                "options": {
                    "required": {
                        "short": "r",
                        "name": "required",
                        "summary": "Test required option",
                        "required": "true"
                    }
                }
            })),
            Context::from_serialize(serde_json::json!({
                "meta": {
                    "name": "usage",
                    "description": "Test simple usage"
                },
                "options": {
                    "multiple": {
                        "short": "m",
                        "name": "multiple",
                        "summary": "Test multiple option",
                        "multiple": "true"
                    }
                }
            })),
            Context::from_serialize(serde_json::json!({
                "meta": {
                    "name": "usage",
                    "description": "Test simple usage"
                },
                "options": {
                    "default": {
                        "short": "d",
                        "name": "default",
                        "summary": "Test default option",
                        "default": "foo"
                    }
                }
            })),
            Context::from_serialize(serde_json::json!({
                "meta": {
                    "name": "usage",
                    "description": "Test simple usage"
                },
                "options": {
                    "choices": {
                        "short": "c",
                        "name": "choices",
                        "summary": "Test option with choices",
                        "choices": ["foo", "bar", "baz"]
                    }
                }
            })),
            Context::from_serialize(serde_json::json!({
                "meta": {
                    "name": "usage",
                    "description": "Test simple usage"
                },
                "options": {
                    "required": {
                        "short": "rm",
                        "name": "required-multiple",
                        "summary": "Test required and multiple option",
                        "required": "true",
                        "multiple": "true"
                    }
                }
            })),
            Context::from_serialize(serde_json::json!({
                "meta": {
                    "name": "usage",
                    "description": "Test simple usage"
                },
                "options": {
                    "default": {
                        "short": "dm",
                        "name": "default-multiple",
                        "summary": "Test default and multiple option",
                        "default": "foo",
                        "multiple": "true"
                    }
                }
            })),
            Context::from_serialize(serde_json::json!({
                "meta": {
                    "name": "usage",
                    "description": "Test simple usage"
                },
                "options": {
                    "choices": {
                        "short": "cm",
                        "name": "choices-multiple",
                        "summary": "Test option with choices and multiple",
                        "choices": ["foo", "bar", "baz"],
                        "multiple": "true"
                    }
                }
            })),
            Context::from_serialize(serde_json::json!({
                "meta": {
                    "name": "usage",
                    "description": "Test simple usage"
                },
                "options": {
                    "all": {
                        "short": "a",
                        "name": "all",
                        "summary": "Test option with all attributes",
                        "choices": ["foo", "bar", "baz"],
                        "multiple": "true",
                        "required": "true",
                        "default": "foo"
                    }
                }
            })),
        ];

        for object in objects {
            let output =
                match TEMPLATES.render("usage.tera", &object.expect("Can't create JSON object")) {
                    Ok(o) => o,
                    Err(e) => {
                        log::error!("Parsing error(s): {}", e);
                        ::std::process::exit(1);
                    }
                };

            println!("output: {:#?}", output);

            insta::assert_snapshot!(output)
        }
    }
}
