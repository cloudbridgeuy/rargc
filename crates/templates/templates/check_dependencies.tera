{%- if dep and dep | length > 0 -%}
  {%- for d in dep %}
  for dependency in {{ d["list"] | join(sep=" ")  }}; do
    if ! command -v $dependency >/dev/null 2>&1; then
      printf "missing dependency: $dependency\n" >&2
      printf "\e[31m%s\e[33m%s\e[31m\e[0m\n\n" "Missing dependency: " "$dependency" >&2
      {%- if dep["message"] %}
      printf "{{ dep["message"] }}\n" >&2
      {%- endif %}
      exit 1
    fi
  done
  {%- endfor %}
{%- endif %}
