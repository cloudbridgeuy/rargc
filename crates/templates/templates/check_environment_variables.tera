{%- set _envs = false %}
{%- if command["envs"] %}
  {%- if envs %}
    {%- set _envs = envs | merge(with=command["envs"]) %}
  {%- else %}
    {%- set _envs = command["envs"] %}
  {%- endif %}
{%- elif envs %}
  {%- set _envs = envs %}
{%- endif %}
{%- if _envs and _envs | length > 0 %}
  {%- for name, env in _envs %}
    {%- if env["option"] %}
  if [[ -n "${% raw %}{{% endraw %}{{ name }}:-}" ]]; then
    args['{{ env["option"] }}']="${% raw %}{{% endraw %}{{ name }}:-}"
  fi
    {%- elif env["required"] %}
  if [[ -z "${% raw %}{{% endraw %}{{ name }}:-}" ]]; then
    printf "\e[31m%s\e[33m%s\e[31m\e[0m\n\n" "Missing required environment variable: " "{{ name }}" >&2
      {%- if command %}
    {{ command["name"] }}_usage >&2
      {%- else %}
    usage >&2
      {%- endif %}
    exit 1
  fi
    {%- endif %}
  {%- endfor %}
{%- endif %}
