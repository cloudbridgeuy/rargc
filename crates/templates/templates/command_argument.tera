{% if command["short"] -%}-{{ command["short"] }} | {% endif -%}--{{ name }})
        action="{{ name }}"
        shift
        {{ name }}_parse_arguments "$@"
        shift $#
        ;;
