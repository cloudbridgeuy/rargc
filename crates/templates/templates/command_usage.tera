{%- set name = name | default(value="script") -%}

{%- set_global required = [] -%}
{%- if options -%}
{%- for key, option in options -%}
  {%- if option["required"] -%}
    {%- set_global required = required | concat(with=option) -%}
  {%- endif -%}
{%- endfor -%}
{%- endif -%}

{{ command["name"] }}_usage() {
  {% if command["description"] -%}
  printf "{{ command["description"] }}\n"
  {% endif -%}

  {% if command["help"] -%}
  printf "{{ command["help"] }}\n"
  {% endif -%}

  printf "\n\033[4m%s\033[0m\n" "Usage:"

  printf "  {{ command["name"] }} {% for option in required %}{% if option["short"] %}-{{ option["short"]}}|{% endif %}--{{ option["name"] }} <{{ option["name"] | upper }}> {% endfor %}[OPTIONS]\n"
  printf "  {{ command["name"] }} -h|--help\n"

  printf "\n\033[4m%s\033[0m\n" "Options:"
  {%- if command["options"] or command["flags"] %}
  {%- if command["options"] %}
  {%- for key, option in command["options"] %}
  printf "  {% if option["short"] -%}-{{ option["short"] }} {% endif -%}--{{ key }} {% if option["required"] %}<{{ key | upper }}>{% else %}[<{{ key | upper }}>]{% endif %}\n"
  printf "    {{ option["summary"] }}\n"
  {%- if option["multiple"] or option["default"] or option["choices"] %}
  printf "    [
    {%- if option["multiple"] -%}
      @multiple
    {%- endif -%}
    {%- if option["default"] -%}
      {%- if option["multiple"] -%}, {% endif -%}@default {{ option["default"] }}
    {%- endif -%}
    {%- if option["choices"] -%}
      {%- if option["multiple"] or option["default"] -%}, {% endif -%}@choices {{ option["choices"] | join(sep=", ") }}
    {%- endif -%}]\n"
  {%- endif -%}
  {%- endfor -%}
  {%- endif %}
  {%- endif %}
  printf "  -h --help\n"
  printf "    Print help\n"

  {%- if options or flags %}
  printf "\n\033[4m%s\033[0m\n" "Global Options:"
  {%- if options %}
  {%- for key, option in options %}
  printf "  {% if option["short"] -%}-{{ option["short"] }} {% endif -%}--{{ key }} {% if option["required"] %}<{{ key | upper }}>{% else %}[<{{ key | upper }}>]{% endif %}\n"
  printf "    {{ option["summary"] }}\n"
  {%- if option["multiple"] or option["default"] or option["choices"] %}
  printf "    [
    {%- if option["multiple"] -%}
      @multiple
    {%- endif -%}
    {%- if option["default"] -%}
      {%- if option["multiple"] -%}, {% endif -%}@default {{ option["default"] }}
    {%- endif -%}
    {%- if option["choices"] -%}
      {%- if option["multiple"] or option["default"] -%}, {% endif -%}@choices {{ option["choices"] | join(sep=", ") }}
    {%- endif -%}]\n"
  {%- endif -%}
  {%- endfor -%}
  {%- endif %}

  {%- if flags %}
  {%- for key, flag in flags %}
  printf "  {% if flag["short"] -%}-{{ flag["short"] }} {% endif -%}--{{ key }}\n"
  printf "    {{ flag["summary"] }}\n"
  {%- endfor %}
  {%- endif %}
  {%- endif %}
}
