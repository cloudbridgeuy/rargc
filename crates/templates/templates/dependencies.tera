# Check global dependencies
check_global_dependencies() {
