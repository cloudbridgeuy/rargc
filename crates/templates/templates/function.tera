{%- set_global required = [] -%}
{%- set_global defaults = [] -%}
{%- set_global choices = [] -%}
{%- set_global _any = false %}
{%- if options -%}
  {%- if command["options"] %}
    {%- for key, option in options | merge(with=command["options"]) -%}
      {%- if option["required"] -%}
        {%- set_global required = required | concat(with=option) -%}
      {%- endif -%}
      {%- if option["default"] -%}
        {%- set_global defaults = defaults | concat(with=option) -%}
      {%- endif -%}
      {% if option["choices"] -%}
        {%- set_global choices = choices | concat(with=option) -%}
      {%- endif -%}
    {%- endfor -%}
  {%- else %}
    {%- for key, option in options -%}
      {%- if option["required"] -%}
        {%- set_global required = required | concat(with=option) -%}
      {%- endif -%}
      {%- if option["default"] -%}
        {%- set_global defaults = defaults | concat(with=option) -%}
      {%- endif -%}
      {% if option["choices"] -%}
        {%- set_global choices = choices | concat(with=option) -%}
      {%- endif -%}
    {%- endfor -%}
  {%- endif %}
{%- else -%}
  {%- if command["options"] -%}
    {%- for key, option in command["options"] -%}
      {%- if option["required"] -%}
        {%- set_global required = required | concat(with=option) -%}
      {%- endif -%}
      {%- if option["default"] -%}
        {%- set_global defaults = defaults | concat(with=option) -%}
      {%- endif -%}
      {% if option["choices"] -%}
        {%- set_global choices = choices | concat(with=option) -%}
      {%- endif -%}
    {%- endfor -%}
  {%- endif -%}
{%- endif -%}
{%- if command["positional_arguments"] -%}
  {%- for arg in command["positional_arguments"] -%}
    {%- if arg["required"] -%}
      {%- set_global required = required | concat(with=arg) -%}
    {%- endif -%}
    {%- if arg["default"] -%}
      {%- set_global defaults = defaults | concat(with=arg) -%}
    {%- endif -%}
    {% if arg["choices"] -%}
      {%- set_global choices = choices | concat(with=arg) -%}
    {%- endif -%}
  {%- endfor -%}
{%- endif -%}
{%- set _envs = false %}
{%- if command["envs"] %}
  {%- if envs %}
    {%- set _envs = envs | merge(with=command["envs"]) %}
  {%- else %}
    {%- set _envs = command["envs"] %}
  {%- endif %}
{%- elif envs %}
  {%- set _envs = envs %}
{%- endif %}
{%- if any %}
  {%- if command["any"] %}
    {%- set _any = command["any"] %}
  {%- else %}
    {%- set _any = any %}
  {%- endif %}
{%- elif command["any"] %}
  {%- set _any = command["any"] %}
{%- endif %}
{%- set opts = false %}
{%- set fls = false %}
{%- set args = false %}
{%- if options %}
  {%- if command["options"] %}
    {%- set opts = options | merge(with=command["options"]) %}
  {%- else %}
    {%- set opts = options %}
  {%- endif %}
{%- else %}
  {%- if command["options"] %}
    {%- set opts = command["options"] %}
  {%- endif %}
{%- endif %}
{%- if flags %}
  {%- if command["flags"] %}
    {%- set fls = flags | merge(with=command["flags"]) %}
  {%- else %}
    {%- set fls = flags %}
  {%- endif %}
{%- else %}
  {%- if command["flags"] %}
    {%- set fls = command["flags"] %}
  {%- endif %}
{%- endif %}
{%- if positional_arguments %}
  {%- if command["positional_arguments"] %}
    {%- set args = command["positional_arguments"] %}
  {%- else %}
    {%- set args = positional_arguments %}
  {%- endif %}
{%- else %}
  {%- if command["positional_arguments"] %}
    {%- set args = command["positional_arguments"] %}
  {%- endif %}
{%- endif %}

{%- if command["description"] -%}
# {{ command["description"] }}
{% endif -%}
{% if command["help"] -%}
# {{ command["help"] }}
{% endif -%}
{{ command["name"] }}() {
{%- if opts %}
  {%- for name, option in opts %}
    {%- if option["multiple"] %}
  declare -a rargs_{{ name | replace(from="-", to="_") }}
    {%- endif %}
  {%- endfor %}
{%- endif %}

{%- if args %}
  {%- for arg in args %}
    {%- if arg["multiple"] %}
  declare -a rargs_{{ name | replace(from="-", to="_") }}
    {%- endif %}
  {%- endfor %}
{%- endif %}

{%- if _envs and _envs | length > 0 %}
  # Parse environment variables
  {% include "check_environment_variables.tera" %}
{%- endif %}
  # Parse command arguments
  parse_{{ command["name"] }}_arguments "$@"
{% if rules -%}
  {%- if rules is containing("no-first-option-help") and rules is not containing("custom-usage") %}
  # Rule `no-first-option-help`: Render the global or command usage if the `-h|--help` option is
  #                              is provided anywhere on the command, not just as the first option.
  #                              Handling individual functions case by case.
  if [[ -n "$rargs_help" ]]; then
    {{ command["name"] }}_usage
    exit 0
  fi
  {%- endif -%}
{%- endif -%}
{%- if command["dep"] and command["dep"] | length > 0 -%}
  {% set dep = command["dep"] %}
  # Check dependencies
  {%- include "check_dependencies.tera" %}
{%- endif %}
{%- if defaults | length > 0 %}
  {% for option in defaults %}
    {% if option["multiple"] %}
  if [[ "${% raw %}{{% endraw %}{% raw %}#{% endraw %}rargs_{{ option["name"] | replace(from="-", to="_") }}[@]}" -eq 0 ]]; then
    rargs_{{ option["name"] | replace(from="-", to="_") }}=({{ option["default"] }})
  fi
    {% else %}
  if [[ -z "$rargs_{{ option["name"] | replace(from="-", to="_") }}" ]]; then
    rargs_{{ option["name"] | replace(from="-", to="_") }}="{{ option["default"] }}"
  fi
    {% endif %}
  {%- endfor %}
{%- endif %}
{%- if choices | length > 0 %}
  {% for option in choices %}
  if [[ -n "$rargs_{{ option["name"] | replace(from="-", to="_") }}" ]]; then
    {%- if option["multiple"] %}
    for value in "${rargs_{{ option["name"] | replace(from="-", to="_") }}[@]}"; do
      if [[ ! "({{ option["choices"] | join(sep=" ") }})" =~ ${value} ]]; then
        printf "\e[31m%s\e[33m%s\e[31m%s\e[33m%s\e[31m\e[0m\n\n" "Invalid option for " "{{ option["name"] }}" ": " "${value}" >&2
        {{ command["name"] }}_usage >&2
        exit 1
      fi
    done
    {%- else %}
    if [[ ! "({{ option["choices"] | join(sep=" ") }})" =~ $rargs_{{ option["name"] | replace(from="-", to="_") }} ]]; then
      printf "\e[31m%s\e[33m%s\e[31m%s\e[33m%s\e[31m\e[0m\n\n" "Invalid option for " "{{ option["name"] }}" ": " "$rargs_{{ option["name"] | replace(from="-", to="_") }}" >&2
      {{ command["name"] }}_usage >&2
      exit 1
    fi
    {%- endif %}
  fi
  {%- endfor %}
{%- endif %}
{%- if _any and _any["required"] %}
  if [[ "{%- raw %}${{%- endraw %}#other_args[@]}" == "0" ]]; then
    printf "\e[31m%s\e[0m\n\n" "Missing required additional argument" >&2
    {{ command["name"] }}_usage >&2
    {%- raw %}
    exit 1
  fi
  {%- endraw %}
{%- endif %}
{%- if required | length > 0 %}
  {% for option in required %}
    {%- if option["multiple"] %}
  if [[ "{% raw %}${{%endraw %}#rargs_{{ option["name"] | replace(from="-", to="_") }}[@]}" == 0 ]]; then
    {%- else %}
  if [[ -z "$rargs_{{ option["name"] | replace(from="-", to="_") }}" ]]; then
    {%- endif %}
    printf "\e[31m%s\e[33m%s\e[31m\e[0m\n\n" "Missing required option: " "{{ option["name"] }}" >&2
    {{ command["name"] }}_usage >&2
    exit 1
  fi
  {%- endfor %}
{%- endif %}
{%- if command["subcommand"] %}
  local sub="{{ command["subcommand"] }}"
{%- endif %}
{%- if command["lines"] %}
  {%- if command["lines"] | length > 0 %}
    {%- for line in command["lines"] %}
{{ line }}
    {%- endfor %}
  {%- endif %}
{%- else %}
  echo "No implementation body found for function {{ command["name"] }}"
{%- endif %}
}
