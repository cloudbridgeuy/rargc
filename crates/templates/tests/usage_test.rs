use tera::Context;
use test_log::test;

use templates::TEMPLATES;
use utils::test_template;

#[test]
fn test_render() {
    test_template!(
        "usage.tera",
        "Typical Usage",
        serde_json::json!({
            "name": "usage",
            "description": "Test simple usage",
            "options": {
                "required": {
                    "short": "r",
                    "name": "required",
                    "description": "Test required option",
                    "required": "true"
                }
            }
        })
    );
    test_template!(
        "usage.tera",
        "Usage with only name and description",
        serde_json::json!({
            "name": "usage",
            "description": "Test simple usage",
        })
    );
    test_template!(
        "usage.tera",
        "Usage with single option",
        serde_json::json!({
            "name": "usage",
            "description": "Test simple usage",
            "options": {
                "foo": {
                    "short": "f",
                    "name": "foo",
                    "description": "Test foo option"
                }
            }
        })
    );
    test_template!(
        "usage.tera",
        "Usage with required option",
        serde_json::json!({
            "name": "usage",
            "description": "Test simple usage",
            "options": {
                "required": {
                    "short": "r",
                    "name": "required",
                    "description": "Test required option",
                    "required": "true"
                }
            }
        })
    );
    test_template!(
        "usage.tera",
        "Usage with multiple option",
        serde_json::json!({
            "name": "usage",
            "description": "Test simple usage",
            "options": {
                "multiple": {
                    "short": "m",
                    "name": "multiple",
                    "description": "Test multiple option",
                    "multiple": "true"
                }
            }
        })
    );
    test_template!(
        "usage.tera",
        "Usage with default option",
        serde_json::json!({
            "name": "usage",
            "description": "Test simple usage",
            "options": {
                "default": {
                    "short": "d",
                    "name": "default",
                    "description": "Test default option",
                    "default": "foo"
                }
            }
        })
    );
    test_template!(
        "usage.tera",
        "Usage with choices option",
        serde_json::json!({
            "name": "usage",
            "description": "Test simple usage",
            "options": {
                "choices": {
                    "short": "c",
                    "name": "choices",
                    "description": "Test choices option",
                    "choices": ["foo", "bar", "baz"]
                }
            }
        })
    );
    test_template!(
        "usage.tera",
        "Usage with required and multiple option",
        serde_json::json!({
            "name": "usage",
            "description": "Test simple usage",
            "options": {
                "required": {
                    "short": "rm",
                    "name": "required-multiple",
                    "description": "Test required and multiple option",
                    "required": "true",
                    "multiple": "true"
                }
            }
        })
    );
    test_template!(
        "usage.tera",
        "Usage with default and multiple option",
        serde_json::json!({
            "name": "usage",
            "description": "Test simple usage",
            "options": {
                "default": {
                    "short": "dm",
                    "name": "default-multiple",
                    "description": "Test default and multiple option",
                    "default": "foo",
                    "multiple": "true"
                }
            }
        })
    );
    test_template!(
        "usage.tera",
        "Usage with options and flags",
        serde_json::json!({
            "name": "usage",
            "description": "Test simple usage",
            "flags": {
                "verbose": {
                    "description": "Test verbose flag"
                },
                "short": {
                    "short": "s",
                    "description": "Test short flag"
                }
            },
            "options": {
                "choices": {
                    "short": "cm",
                    "name": "choices-multiple",
                    "description": "Test option with choices and multiple",
                    "choices": ["foo", "bar", "baz"],
                    "multiple": "true"
                }
            }
        })
    );
    test_template!(
        "usage.tera",
        "Usage with commands",
        serde_json::json!({
            "name": "usage",
            "description": "Test simple usage",
            "commands": {
                "foo": {
                    "description": "Test foo command"
                }
            }
        })
    );
    test_template!(
        "usage.tera",
        "Usage with commands and flags",
        serde_json::json!({
            "name": "usage",
            "description": "Test simple usage",
            "flags": {
                "verbose": {
                    "description": "Test verbose flag"
                }
            },
            "commands": {
                "foo": {
                    "description": "Test foo command"
                }
            }
        })
    );
    test_template!(
        "usage.tera",
        "Usage with commands and options",
        serde_json::json!({
            "name": "usage",
            "description": "Test simple usage",
            "options": {
                "choices": {
                    "short": "cm",
                    "name": "choices-multiple",
                    "description": "Test option with choices and multiple",
                    "choices": ["foo", "bar", "baz"],
                    "multiple": "true"
                }
            },
            "commands": {
                "foo": {
                    "description": "Test foo command"
                }
            }
        })
    );
    test_template!(
        "usage.tera",
        "Usage with commands, flags and options",
        serde_json::json!({
            "name": "usage",
            "description": "Test simple usage",
            "flags": {
                "verbose": {
                    "description": "Test verbose flag"
                }
            },
            "options": {
                "choices": {
                    "short": "cm",
                    "name": "choices-multiple",
                    "description": "Test option with choices and multiple",
                    "choices": ["foo", "bar", "baz"],
                    "multiple": "true"
                }
            },
            "commands": {
                "foo": {
                    "description": "Test foo command"
                }
            }
        })
    );
    test_template!(
        "usage.tera",
        "Usage with extra help message",
        serde_json::json!({
            "name": "usage",
            "description": "Test simple usage",
            "name": "foo".to_string(),
            "options": {
                "required": {
                    "short": "r",
                    "name": "required",
                    "description": "Test required option",
                    "required": "true"
                }
            }
        })
    );
    test_template!(
        "usage.tera",
        "Usage with option with all flags",
        serde_json::json!({
            "name": "usage",
            "description": "Test simple usage",
            "default": "foo",
            "flags": {
                "verbose": {
                    "description": "Test verbose flag"
                }
            },
            "commands": {
                "foo": {
                    "description": "Test foo command",
                    "help": "Some additional help message"
                }
            },
            "options": {
                "all": {
                    "short": "a",
                    "name": "all",
                    "description": "Test option with all attributes",
                    "choices": ["foo", "bar", "baz"],
                    "multiple": "true",
                    "required": "true",
                    "default": "foo"
                }
            }
        })
    );
    test_template!(
        "usage.tera",
        "Usage with multiple commands with default",
        serde_json::json!({
            "name": "usage",
            "description": "Test simple usage",
            "default": "foo",
            "flags": {
                "verbose": {
                    "description": "Test verbose flag"
                }
            },
            "commands": {
                "foo": {
                    "description": "Test foo command",
                    "help": "Some additional help message"
                },
                "bar": {
                    "description": "Test bar command",
                    "help": "Some additional help message"
                },
                "baz": {
                    "description": "Test baz command",
                    "help": "Some additional help message"
                }
            },
        })
    );
    test_template!(
        "usage.tera",
        "Usage with multiple commands",
        serde_json::json!({
            "name": "usage",
            "description": "Test simple usage",
            "flags": {
                "verbose": {
                    "description": "Test verbose flag"
                }
            },
            "commands": {
                "foo": {
                    "description": "Test foo command",
                    "help": "Some additional help message"
                },
                "bar": {
                    "description": "Test bar command",
                    "help": "Some additional help message"
                },
                "baz": {
                    "description": "Test baz command",
                    "help": "Some additional help message"
                }
            },
        })
    );
    test_template!(
        "usage.tera",
        "Usage with examples",
        serde_json::json!({
            "name": "usage",
            "description": "Test simple usage",
            "examples": [{
                "description": "Verbose mode",
                "command": "-v|--verbose"
            }, {
                "description": "Call command foo in verbose mode",
                "command": "-v|--verbose foo",
            }],
            "flags": {
                "verbose": {
                    "short": "v",
                    "description": "Test verbose flag"
                }
            },
            "options": {
                "choices": {
                    "short": "cm",
                    "name": "choices-multiple",
                    "description": "Test option with choices and multiple",
                    "choices": ["foo", "bar", "baz"],
                    "multiple": "true"
                }
            },
            "commands": {
                "foo": {
                    "description": "Test foo command",
                    "help": "Some additional help message"
                },
                "bar": {
                    "description": "Test bar command",
                    "help": "Some additional help message"
                },
                "baz": {
                    "description": "Test baz command",
                    "help": "Some additional help message"
                }
            },
        })
    );
    test_template!(
        "usage.tera",
        "Usage with positional arguments",
        serde_json::json!({
            "name": "usage",
            "description": "Test usage with positional arguments",
            "positional_arguments": [{
                "name": "foo",
                "description": "Test foo command"
            }, {
                "name": "bar",
                "description": "Test bar command"
            }]
        })
    );

    test_template!(
        "usage.tera",
        "Usage with positional arguments",
        serde_json::json!({
            "name": "usage",
            "description": "Test usage with positional arguments, one of them required",
            "positional_arguments": [{
                "name": "foo",
                "description": "Test foo command",
                "required": true
            }, {
                "name": "bar",
                "description": "Test bar command"
            }]
        })
    );

    test_template!(
        "usage.tera",
        "Usage with a global option that uses -h as short",
        serde_json::json!({
            "name": "usage",
            "description": "Test usage with a global option that uses -h as short",
            "options": {
                "host": {
                    "short": "h",
                    "name": "host",
                    "description": "Test host option"
                }
            }
        })
    );

    test_template!(
        "usage.tera",
        "Usage with a global flag that uses -h as short",
        serde_json::json!({
            "name": "usage",
            "description": "Test usage with a global option that uses -h as short",
            "flags": {
                "host": {
                    "short": "h",
                    "name": "host",
                    "description": "Test host option"
                }
            }
        })
    );

    test_template!(
        "usage.tera",
        "Usage with a global option that uses -v as short",
        serde_json::json!({
            "name": "usage",
            "version": "0.0.1",
            "description": "Test usage with a global option that uses -v as short",
            "options": {
                "verbose": {
                    "short": "v",
                    "name": "host",
                    "description": "Test verbose option"
                }
            }
        })
    );

    test_template!(
        "usage.tera",
        "Usage with a global flag that uses -v as short",
        serde_json::json!({
            "name": "usage",
            "version": "0.0.1",
            "description": "Test usage with a global option that uses -v as short",
            "flags": {
                "verbose": {
                    "short": "v",
                    "name": "host",
                    "description": "Test verbose option"
                }
            }
        })
    );

    test_template!(
        "usage.tera",
        "Usage with option with all flags and value notations",
        serde_json::json!({
            "name": "usage",
            "description": "Test simple usage",
            "default": "foo",
            "flags": {
                "verbose": {
                    "description": "Test verbose flag"
                }
            },
            "commands": {
                "foo": {
                    "description": "Test foo command",
                    "help": "Some additional help message"
                }
            },
            "positional_arguments": [{
                "name": "some",
                "description": "Test argument with all attributes",
                "choices": ["foo", "bar", "baz"],
                "multiple": "true",
                "required": "true",
                "default": "foo",
                "value_notation": "ELSE"
            }],
            "options": {
                "all": {
                    "short": "a",
                    "name": "all",
                    "description": "Test option with all attributes",
                    "choices": ["foo", "bar", "baz"],
                    "multiple": "true",
                    "required": "true",
                    "default": "foo",
                    "value_notation": "ELSE"
                }
            }
        })
    );

    test_template!(
        "usage.tera",
        "Usage with option with all flags and value notations",
        serde_json::json!({
            "name": "usage",
            "description": "Test simple usage",
            "default": "foo",
            "flags": {
                "verbose": {
                    "description": "Test verbose flag"
                }
            },
            "commands": {
                "foo": {
                    "description": "Test foo command",
                    "help": "Some additional help message",
                    "positional_arguments": [{
                        "name": "some",
                        "description": "Test argument with all attributes",
                        "choices": ["foo", "bar", "baz"],
                        "multiple": "true",
                        "required": "true",
                        "default": "foo",
                        "value_notation": "ELSE"
                    }],
                    "options": {
                        "all": {
                            "short": "a",
                            "name": "all",
                            "description": "Test option with all attributes",
                            "choices": ["foo", "bar", "baz"],
                            "multiple": "true",
                            "required": "true",
                            "default": "foo",
                            "value_notation": "ELSE"
                        }
                    }
                }
            },
        })
    );
}
