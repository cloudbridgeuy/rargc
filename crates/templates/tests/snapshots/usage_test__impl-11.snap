---
source: crates/templates/tests/usage_test.rs
description: Usage with commands
expression: output
---
usage() {
  printf "Test simple usage\n"
  printf "\n\033[4m%s\033[0m\n" "Usage:"
  printf "  usage [OPTIONS] [COMMAND] [COMMAND_OPTIONS]\n"

  printf "  usage -h|--help\n"
  printf "\n\033[4m%s\033[0m\n" "Commands:"
  printf "  foo\t\n"

  printf "\n\033[4m%s\033[0m\n" "Options:"
  printf "  -h --help\n"
  printf "    Print help\n"
}

